examples/directchat.nim

```diff
@@ -63,143 +97,97 @@ proc dialPeer(p: ChatProto, address: string) {.async.} =
     wireAddr = ip4Addr & tcpAddr
 
   echo &"dialing peer: {multiAddr}"
-  p.conn = await p.switch.dial(remotePeer, @[wireAddr], ChatCodec)
-  p.connected = true
-  asyncSpawn p.readAndPrint()
+  asyncSpawn c.handlePeer(await c.switch.dial(remotePeer, @[wireAddr], ChatCodec))
 
-proc writeAndPrint(p: ChatProto) {.async.} =
+proc readLoop(c: Chat) {.async.} =
   while true:
-    if not p.connected:
+    if not c.connected:
       echo "type an address or wait for a connection:"
       echo "type /[help|?] for help"
 
-    let line = await p.transp.readLine()
-    if line.startsWith("/help") or line.startsWith("/?") or not p.started:
+    c.writePrompt()
+
+    let line = await c.stdinReader.readLine()
+    if line.startsWith("/help") or line.startsWith("/?"):
       echo Help
       continue
 
     if line.startsWith("/disconnect"):
-      echo "Ending current session"
-      if p.connected and p.conn.closed.not:
-        await p.conn.close()
-      p.connected = false
+      c.writeStdout "Ending current session"
+      if c.connected and c.conn.closed.not:
+        await c.conn.close()
+      c.connected = false
     elif line.startsWith("/connect"):
-      if p.connected:
-        var yesno = "N"
-        echo "a session is already in progress, do you want end it [y/N]?"
-        yesno = await p.transp.readLine()
-        if yesno.cmpIgnoreCase("y") == 0:
-          await p.conn.close()
-          p.connected = false
-        elif yesno.cmpIgnoreCase("n") == 0:
-          continue
-        else:
-          echo "unrecognized response"
-          continue
-
-      echo "enter address of remote peer"
-      let address = await p.transp.readLine()
+      c.writeStdout "enter address of remote peer"
+      let address = await c.stdinReader.readLine()
       if address.len > 0:
-        await p.dialPeer(address)
+        await c.dialPeer(address)
 
     elif line.startsWith("/exit"):
-      if p.connected and p.conn.closed.not:
-        await p.conn.close()
-        p.connected = false
+      if c.connected and c.conn.closed.not:
+        await c.conn.close()
+        c.connected = false
 
-      await p.switch.stop()
-      echo "quitting..."
+      await c.switch.stop()
+      c.writeStdout "quitting..."
       quit(0)
     else:
-      if p.connected:
-        await p.conn.writeLp(line)
+      if c.connected:
+        await c.conn.writeLp(line)
       else:
         try:
           if line.startsWith("/") and "p2p" in line:
-            await p.dialPeer(line)
+            await c.dialPeer(line)
         except:
```

Comment:
I think we're now avoiding this style due to multiple language issues. It's probably better to use `exception CatchableError as exc`.

Reply:
Yeah, it's the issue with long running PRs, I would have fixed this in the various cleanups part 1 :)